[ASTImporter] Improved error detection at import

balazske opened the pull request:
First version of the change in ASTImporter to improve import error detection and error information (including One-Definition-Rule).

This change is not complete, the test Analysis/ctu-main.c fails. To have accurate information about import errors, it is needed that the error is stored for `Expr` and `Stmt` too. After this it can be possible to detect what kind of error caused a `FunctionDecl` body import failure, and filter out for unsupported-construct and ODR errors. These are not problems in CTU import, but the import will assert if any other type of error is detected (if this case is possible at all).

The import "logic" in `VisitRecordDecl` was changed to correctly detect name conflict (and only if it is really problem). Still I am almost sure that it is not perfect.

balazske committed:
[ASTImporter] Improved error detection at import.

martong commented on include/clang/AST/ASTImporter.h:
```diff
@@ -253,9 +285,11 @@ namespace clang {
+    /// If the original name is returned the conflict remains unresolved such
+    /// that name lookup for this name is likely to find an ambiguity later.
+    /// This can cause various assertions or problems later
+    /// and should be avoided.
```
Can't we just remove `virtual` from `HandleNameConflict` ? If it is not overriden in any of the subclasses neither in Clang nor in LLDB then why do we need that to be virtual? We could just simply make it a private non-virtual function.

martong commented on include/clang/AST/ASTImporter.h:
```diff
@@ -314,6 +349,10 @@ namespace clang {
     Decl *MapImported(Decl *From, Decl *To);
 
+    /// Mark imported declaration with error.
+    /// Returns \c nullptr
```
Returning with `nullptr` does not make so much sense to me. Why not `void`?

martong replied:
Or the comment is not aligned with the definition.

martong commented on include/clang/AST/ASTImporter.h:
```diff
@@ -327,13 +366,11 @@ namespace clang {
-    void setEncounteredUnsupportedConstruct(bool B) {
-      EncounteredUnsupportedConstruct = B;
-    }
+    llvm::DenseMap<int, unsigned int> ImportDeclErrorCount;
```
Somehow we should indicate that `ImportDeclError` is not going to be erased/cleared during the life-time of the Importer object, however this can be. Perhaps another naming, but at least comments should indicate that.

martong commented on lib/CrossTU/CrossTranslationUnit.cpp:
```diff
@@ -73,6 +73,8 @@ STATISTIC(NumGetCTUSuccess, "The # of getCTUDefinition successfully return the "
                             "requested function's body");
 STATISTIC(NumUnsupportedNodeFound, "The # of imports when the ASTImporter "
                                    "encountered an unsupported AST Node");
+STATISTIC(NumODRErrorFound, "The # of imports when the ASTImporter "
```
"some kind of" sounds a bit bloated to me.

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3840,6 +3843,177 @@ TEST_P(DeclContextTest, removeDeclOfClassTemplateSpecialization) {
+      )", Lang_CXX);
+  auto *FromClass = FirstDeclMatcher<CXXRecordDecl>().match(
+      FromTU, cxxRecordDecl(hasName("X")));
+  auto *ToClass = Import(FromClass, Lang_CXX);
```
Perhaps a better name would be `ImportedClass` since `ToClass` is the class already in the "To" context before the import even started.

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3840,6 +3843,177 @@ TEST_P(DeclContextTest, removeDeclOfClassTemplateSpecialization) {
+  EXPECT_EQ(ToClass, nullptr);
+}
+
+TEST_P(CheckODRAtImport, DISABLED_Function) {
```
Is this disabled because structural eq find them different?

balazske replied:
The "from" function is imported beneath the existing, but this results in invalid code (no overload at return type is allowed). These functions are detected as different by structural equivalence test.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -1363,8 +1382,10 @@ bool ASTNodeImporter::ImportDefinition(RecordDecl *From, RecordDecl *To,
       QualType T = Importer.Import(Base1.getType());
-      if (T.isNull())
+      if (T.isNull()) {
+        Importer.setImportDeclError(From, ImportErrorKind::InheritedDecl);
```
InheritedType?

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -1256,14 +1267,22 @@ void ASTNodeImporter::ImportDeclContext(DeclContext *FromDC, bool ForceImport,
-      Importer.Import(FromM);
+  for (CXXMethodDecl *FromM : From->methods()) {
+    if (FromM->isImplicit()) {
+      if (!Importer.Import(FromM)) {
```
We have to do a similar refactor in case of ClassTemplateSpecializations.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -1417,16 +1438,21 @@ bool ASTNodeImporter::ImportDefinition(EnumDecl *From, EnumDecl *To,
   }
   
   To->startDefinition();
+  // FIXME: completeDefinition at error return cases?
```
I think we can leave that in `started` state in case of error.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2252,8 +2288,7 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
                   D->isAnonymousStructOrUnion()
-                    == FoundDef->isAnonymousStructOrUnion() &&
-                  IsStructuralMatch(D, FoundDef))) {
+                    == FoundDef->isAnonymousStructOrUnion())) {
```
Are you sure it is ok to remove the structural eq check from here? E.g. we should import the implicit methods only if they are equivalent ...

balazske replied:
The same check should be done at line 2279.

martong reviewed (commented):
Nice work, this is a big change and I think the direction is good.
If we are changing so many callee sites of `Import`, could we change `Import()` to be nodiscard?

gamesh411 commented:
Can one of the admins verify this patch?

balazske committed:
Second version

balazske commented:
New commit added:
The import error is stored in a variable and can be set from any import function. If a Decl import fails it is read from this and set for the Decl.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2286,18 +2324,18 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
-            continue;
-
-          if (IsFriendTemplate)
+              !IsStructuralMatch(D, FoundRecord)) {
```
Will this condition be ever true? In line 2312 we have `else if (!D->isCompleteDefinition()) {` as the enclosing condition,  so line 2326 will never be true.

martong commented on include/clang/AST/ASTImporter.h:
```diff
@@ -95,9 +93,28 @@ namespace clang {
+    /// successful but some of its "subnodes" is not successful (for example at
+    /// members of a namespace, some members can be missed because import
+    /// failure but the import of the whole namespace itself still succeeds).
+    llvm::DenseMap<int, unsigned int> ImportDeclErrorCount;
```
Why not `llvm::DenseMap<ImportErrorKind, unsigned int> ImportDeclErrorCount;` ?

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -7203,8 +7197,11 @@ TypeSourceInfo *ASTImporter::Import(TypeSourceInfo *FromTSI) {
-  if (T.isNull())
+  if (T.isNull()) {
+    // Set the error if not set already more specifically.
+    setCurrentImportDeclError(ImportErrorKind::TypeFailed);
```
Perhaps we should have an error kind for all `Import()` functions. In this case it should be `TypeSourceInfoFailed`.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -7182,10 +7263,20 @@ Decl *ASTImporter::Import(Decl *FromD) {
+  } else {
+    auto Error = getImportDeclErrorIfAny(FromD);
+    if (!Error) {
+      setImportDeclError(FromD, *CurrentImportDeclError);
```
We should assert here that `CurrentImportDeclError` is set.

martong reviewed (commented):
I think the logic about how to handle the error states is good.

But I have concerns with `CurrentImportDeclError`. Actually, that is a hack to hold the result/error of the different toplevel `Import()` functions.
And we can store the result only in case of Decls. We have to manually reset this, which seems error prone. How could we generalize this approach if we want to keep track of errors for e.g. Types too? Should we have a `CurrentImportTypeError`?

Seems like we are trying to avoid to pass the error/result in the return value.
I think the code would be much easier to reason about if we would use return values, and we could generalize the approach used with Decls to QualTypes and other nodes more easily.
I am thinking about to change the Import functions like this: `Expr* Import(Expr*)` -> `[[nodiscard]] ErrorOr<Expr*> Import(Expr*)`.
http://llvm.org/doxygen/classllvm_1_1ErrorOr.html#details . We could add `[[nodiscard]]` in a later step. `ErrorOr`  is convertible to bool like a nullptr, so hopefully we don't have to change the callee sites.
This may result an even bigger change, but since we are already deep inside it then I definitely vote for changing the return types of the top level `Import()` functions.

martong commented:
Just one more thought:
We have to change the signature of the top level `Import()` functions only, but we can/must leave intact the signature of the `Visit*` functions. Since we would return from the top level import function with an error if the visitation resulted with a nullptr (as you implemented that in your last commit).

martong commented:
Only one question remained: How to handle the name conflicts?
We could handle them as a simple state in `ASTNodeImporter`, since all top level `Import()` functions of `ASTImporter` has a local object of the node importer and after the `Visit()` call we could query the name conflict state.

martong commented:
`CurrentImportDeclError` seems like a dangerous global state, instead we could have a more localised error state in `ASTNodeImporter` which could be set before all `return nullptr`. But this is just a middle step before finally we change the signature of the `T* VisitT` functions too to `ErrorOr<T> VisitT`, sooner or later we must do that also.

martong commented:
I think, there are only two major cause of import errors:
* name conflict (odr violation)
* unsupported node

The other indirect errors are all derived from these and not really relevant.
Thus, we could simply forward the error via the call chain as a return value (once we changed both the `Import()` and `Visit()` function signatures):
```
auto ImportedOrError = Importer.Import(E);
if (!ImportedOrError)
  return ImportedOrError;
```

balazske commented:
So we should make the big change of change every Import function (not only for `Decl`) to return an `Expected<Decl *>` (or similar)? (`Expected` should be better than `ErrorOr`, but the needed code to convert the error into a `std::error_code` where a `std::error_category` is needed will not be implemented if possible.) Not only the Import functions must be changed, the other utility functions (mostly for template related things and DeclContext import) too.

balazske committed:
Various fixes mainly at test.

balazske committed:
Merge branch 'ctu-clang6' into ODR-failure

balazske committed:
Error check at Import - step 1

Removed previous code (temporary comments left).
Added new error class.
Changed Import of DeclarationName.

balazske commented:
New commit added: First part of implementation of complete error check, it should be done like this for every Import function, and the visit functions should return Expected too.
This is a temporary kind of change, some old comments and code parts are left, and CTU import function is not complete, error kind counters are not implemented (probably not needed).

martong commented on include/clang/AST/ASTImporter.h:
```diff
@@ -36,6 +38,30 @@ namespace clang {
+
+    ErrorKind Error;
+
+    static char ID;
```
Why do we need this `ID`?

balazske replied:
It is used by `ErrorInfo`.

martong commented on lib/AST/ExternalASTMerger.cpp:
```diff
@@ -57,7 +57,14 @@ LookupSameContext(Source<TranslationUnitDecl *> SourceTU, const DeclContext *DC,
+  auto NameOrErr = ReverseImporter.Import(Name);
+  if (auto Err = NameOrErr.takeError()) {
+    // Ignore error.
+    handleAllErrors(std::move(Err), [](const ImportError&) { });
```
Ideally we would do `return Err`, but that would require changing the signature of `LookupSameContext()` too.
I think, it is perfectly fine to change the signature of the `Import` and `Visit` functions first, then later we could address these kind of auxiliary functions too.

martong reviewed (commented):
Looks good, keep on going. :+1:

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -1136,8 +1175,8 @@ bool ASTNodeImporter::ImportDeclParts(NamedDecl *D, DeclContext *&DC,
-  Name = Importer.Import(D->getDeclName());
-  if (D->getDeclName() && !Name)
+  if (auto Err = ImportOrError(Name, D->getDeclName()))
+    // FIXME: handle error
```
We could just return the `Err` here, which means ImportDeclParts should change its return type from `bool` to `Error`.

balazske replied:
It is planned to change these helper import functions to return Error or Expected.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2100,7 +2152,10 @@ Decl *ASTNodeImporter::VisitEnumDecl(EnumDecl *D) {
+    if (auto Err = ImportOrError(
+        SearchName, D->getTypedefNameForAnonDecl()->getDeclName()))
+      // FIXME: handle error
+      return nullptr;
```
Once the visit functions have a return type of `Expected<T>` then this will be `return std::move(Err);` , right?

balazske replied:
Yes, the error should be returned.

balazske commented:
Using exceptions instead of check error and return error would be the more convenient way of doing this. The whole import code will be "full" of these error check and return error statements.

martong commented:
Yes, I agree that exceptions are superior in error handling, however `fno-expections` is an LLVM policy and thus we must comply. Other than that, these error cases may happen often during a CTU analysis and we know that exceptions can be really slow on the throw-catch branch, so this error handling with return values will be more effective I think.

balazske committed:
Error check at Import - step 2

Changed import of SourceRange, and some other things.

martong reviewed (commented):
phase 2 looks good to me :+1:

balazske committed:
Error check at Import - step 3

balazske commented:
Committed "step 3".

The import "helper" functions are updated to return `Error` or `Expected` types. Error handling is added to places where it was omitted.
A new "ImportDeclContext" function was needed. Probably it can be renamed, or the original (old) ImportDeclContext can be renamed.

Error handling is not  implemented, the code compiles but many tests do not pass. This will work only after every import and visit function can return the error object.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -247,42 +254,47 @@ namespace clang {
-    std::tuple<FunctionTemplateDecl *, OptionalTemplateArgsTy>
-    ImportFunctionTemplateWithTemplateArgsFromSpecialization(
-        FunctionDecl *FromFD);
+    Error ImportFunctionTemplateWithTemplateArgsFromSpecialization(
```
`Expected<std::tuple<FunctionTemplateDecl *, OptionalTemplateArgsTy>>` ?

balazske replied:
A `FunctionTemplateDecl *` and a `TemplateArgsTy` are two output values from the function. But if any of these fails to import the whole function will fail (or not?), there is no valid case when there was no import error and one of the output arguments is null (empty optional or null pointer). So optional is not needed at return, and it is more simple to have output parameters like at every other similar function.
There are two kind of import function, one that has a reference output value and returns `Error` and one that returns `Expected`. The first should update something in an existing object, the second type should return something new. (For this reason a tuple is the better solution here, only it was somewhat inconvenient.)

martong replied:
Yeah I think we agree. My understanding is that we should return `Error` in case of functions which previously did not return with any meaningful value, e.g. `void` or `bool` to indicate error. In those cases where the function returns with a  meaningful value we should use `Expected`.

balazske replied:
Still it is hard to decide when to use the `Error` return value and when to use `Expected`. If a function has a vector output value it may be not efficient (?) to return it with Expected. For `SmallVector` the used "size" of the vector is different when `ImportFunctionTemplateWithTemplateArgsFromSpecialization` is called, the parameter vector is not size-dependent but a returned vector has the same size ("size": template parameter for `SmallVector`). Probably it is better to make everything work with output parameters and `Error` return value.

martong replied:
>  If a function has a vector output value it may be not efficient (?) to return it with Expected.

I think, if the cost of returning `T` is 1 unit than the cost of returning `Expected<T>` will be 1 unit too. I mean it should be if the template is written properly.

> For SmallVector the used "size" of the vector is different when ImportFunctionTemplateWithTemplateArgsFromSpecialization is called, the parameter vector is not size-dependent but a returned vector has the same size ("size": template parameter for SmallVector). 

I don't quite understand this. The size template parameter of `SmallVector` should be the same, we use `OptionalTemplateArgsTy`.

> Probably it is better to make everything work with output parameters and Error return value.

I don't think so, I rather think the receipt is very simple: if T is not `void` or `bool` then `T f() ---> Expected<T> f()` , otherwise use `Error`.

balazske replied:
The problem is with `ImportTemplateArguments`, it has currently a vector reference output parameter but it should return this value with `Expected`. This function is used with `TemplateArgsTy` and with other `SmallVector` (with size 2).

balazske replied:
And the `ImportDeclParts` function should be changed to return a big tuple inside an `Expected`?

martong replied:
TLDR, we can have the output params, it is OK.


Just some other thoughts:

> And the ImportDeclParts function should be changed to return a big tuple inside an Expected?

Okay, `ImportDeclParts` is already taking output parameters.
But for me it would be easier to understand the code if we never used output parameters, especially not with reference types. Because it is very hard to see on the callee site that we pass an output parameter. That is why I used previously tuples to return multiple values.

So, I'd rather have `ImportDeclParts` returning with a tuple. But then we should change also `GetImportedOrCreateDecl` and who knows what else.

martong replied:
> The problem is with ImportTemplateArguments, it has currently a vector reference output parameter but it should return this value with Expected. This function is used with TemplateArgsTy and with other SmallVector (with size 2).

We can use `SmallVectorImpl` in such cases:
In APIs that don’t care about the “small size” (most?), prefer to use the SmallVectorImpl<T> class, which is basically just the “vector header” (and methods) without the elements allocated after it. Note that SmallVector<T, N> inherits from SmallVectorImpl<T> so the conversion is implicit and costs nothing.

balazske replied:
I do not want refactoring now, just "mechanical" changes. The `ImportFunctionTemplateWithTemplateArgsFromSpecialization` will return tuple in Expected, and what returned `void` or `bool` will return `Error`.

martong replied:
If we have an `Expected<T>` as a result type, then there is no way to acquire the `T` if there was an error. However, when we have output parameters and `Error`, then even if there was an error some output params could have been set ... and those can be reused even after the return with the `Error`. If error handling is not proper on the callee site then we may continue with stale values, which is not possible if we use `Expected<T>`.

And I think the above is quite a strong argument against output params. Now I changed my mind, could we just remove all output params?

martong replied:
> And I think the above is quite a strong argument against output params. Now I changed my mind, could we just remove all output params?

Or at least maybe we should just not add new functions which uses output params. (Maybe later we could change existing functions)

martong replied:
> I do not want refactoring now, just "mechanical" changes. The ImportFunctionTemplateWithTemplateArgsFromSpecialization will return tuple in Expected, and what returned void or bool will return Error.

Ok, that's fine for me.

balazske replied:
Another thing to refactor: Make the input parameters of import functions `const`.

martong replied:
:+1:

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4772,14 +4797,14 @@ Decl *ASTNodeImporter::VisitClassTemplateSpecializationDecl(
-      TemplateParameterList *ToTPList = ImportTemplateParameterList(
-            PartialSpec->getTemplateParameters());
-      if (!ToTPList && PartialSpec->getTemplateParameters())
+      auto ToTPListOrErr = ImportTemplateParameterList(
```
Perhaps instead of `auto` it is better to explicitly say that this is an `Error`. (Or is it an `Expected` ? :) )
I'd use `auto` only if the same statement contains the explicit type like in case of `cast<>`.

balazske replied:
I wanted to use `Error` if `Error` is returned, but `auto` if an `Expected<something-that-maybe-long>` is returned to make the code simpler.

martong replied:
Ok.

martong reviewed (commented):
Phase 3 looks good to me. Just found some minor `auto` related style thing.

balazske committed:
Changed return value of ImportFunctionTemplateWithTemplateArgsFromSpecialization.

balazske committed:
step 4 - import of SourceRange, SourceLocation

balazske commented:
Next part: Import of `SourceRange` and `SourceLocation` is changed to use Expected. This resulted in very big amount of changes. Many of the SourceLocation import calls were inside a function call, this was moved out with a new variable for the imported value. The imported values are returned from `Import` that returns an `Expected<T>`, the `Expected<T>` type is used (not `T`) (`auto` is used to make the code more readable and simple). These values are called `XxxOrErr` like in the LLVM documentation (for Expected). After checked for error, the * operator can be used to get the value. In this way the previous `ImportOrError` function is not needed, probably it can be even removed, because we do not want to use output parameters. After every import is changed, many of the parameters to `GetImportedOrCreateDecl` will be a "XxxOrErr" type of value.

martong reviewed (commented):
phase 4 looks really good to me! :+1:

balazske committed:
Changed import of TemplateName.

balazske committed:
Changed import of NestedNameSpecifier *, and CXXBaseSpecifier, CXXCtorInitializer.

Renamed ImportOrError to importInto.
Removed some uses of ImportOrError.

balazske committed:
Changed import of TypeSourceInfo * .

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -522,6 +522,13 @@ namespace clang {
     }
 
+    template<typename InContainerTy, typename OutContainerTy>
+    Error ImportContainerCheckedNew(
```
Could we add a `FIXME`?

martong commented:
:+1:  The last 3 commits look good, keep on pushing :)

balazske committed:
Changed import of QualType.

balazske commented:
New commits uploaded, now it is ready until QualType.
The "discErrorType" is used to take the error and handle it (ignore?), so it do not remain unhandled. This can be used at every place when an Error needs to be returned. In this way it is possible to get working code without change the import of Decl *, Stmt *, Expr * (these are the next steps).

balazske committed:
Changed type visitor function to use Expected.

martong reviewed (commented):
Looks good.

balazske commented:
Question: If the return type of `VisitXxxStmt` will be `Expected<Stmt *>`, what should be the return type of `VisitXxxExpr`? Can this be `Expected<Expr *>`, or `Expected<Stmt *>` (I think this should work). The Visit is only called by the Import function with `Stmt *` directly, the return type of VisitXxxExpr does not matter (why is it `Expr *` instead of `Stmt *`?).

balazske commented:
What about an import template like this?
```
template <class FromT, class ... FromTArgs>
Expected<std::tuple<FromT, FromTArgs...>> import(const FromT &From, const FromTArgs &...FromArgs)
```
This requires much rewrite of the visit functions but the code would be shorter and more simple.

martong commented:
> Question: If the return type of VisitXxxStmt will be Expected<Stmt *>, what should be the return type of VisitXxxExpr?

I think VisitXXXExpr should return with Expected<Expr *>. However Expr is derived from Stmt, so it does not really matter, but Expected<Expr *> seems more consistent.

martong commented:
> What about an import template like this?
```
template <class FromT, class ... FromTArgs>
Expected<std::tuple<FromT, FromTArgs...>> import(const FromT &From, const FromTArgs &...FromArgs)
```
> This requires much rewrite of the visit functions but the code would be shorter and more simple.

Could you give an example for the usage of this template? I don't quite understand the intention yet ...

balazske commented:
For example `VisitCaseStmt` would look like this, after every import is error handled:
```
ExpectedStmt ASTNodeImporter::VisitCaseStmt(CaseStmt *S) {
  ExpectedExpr ToLHSOrErr = import(S->getLHS());
  if (!ToLHSOrErr)
    return ToLHSOrErr.takeError();
  ExpectedExpr ToRHSOrErr = import(S->getRHS());
  if (!ToRHSOrErr)
    return ToRHSOrErr.takeError();
  ExpectedStmt ToSubStmtOrErr = import(S->getSubStmt());
  if (!ToSubStmtOrErr)
    return ToSubStmtOrErr.takeError();
  auto ToCaseLocOrErr = Importer.Import(S->getCaseLoc());
  if (!ToCaseLocOrErr)
    return ToCaseLocOrErr.takeError();
  auto ToEllipsisLocOrErr = Importer.Import(S->getEllipsisLoc());
  if (!ToEllipsisLocOrErr)
    return ToEllipsisLocOrErr.takeError();
  auto ToColonLocOrErr = Importer.Import(S->getColonLoc());
  if (!ToColonLocOrErr)
    return ToColonLocOrErr.takeError();
  CaseStmt *ToStmt = new (Importer.getToContext())
      CaseStmt(ToLHS, ToRHS,
               *ToCaseLocOrErr, *ToEllipsisLocOrErr, *ToColonLocOrErr);
  ToStmt->setSubStmt(ToSubStmt);
  return ToStmt;
}
```
With the new template it can be like this:
```
ExpectedStmt ASTNodeImporter::VisitCaseStmt(CaseStmt *S) {
  Expr *ToLHS, *ToRHS;
  Stmt *ToSubStmt;
  SourceLocation ToCaseLoc, ToEllipsisLoc, ToColonLoc;

  auto Imp = import(
      S->getLHS(), S->getRHS(), S->getSubStmt(), S->getCaseLoc(),
      S->getEllipsisLoc(), S->getColonLoc());
  if (!Imp)
    return Imp.takeError();
  std::tie(ToLHS, ToRHS, ToSubStmt, ToCaseLoc, ToEllipsisLoc, ToColonLoc) =
      *Imp;

  CaseStmt *ToStmt = new (Importer.getToContext()) CaseStmt(
      ToLHS, ToRHS, ToCaseLoc, ToEllipsisLoc, ToColonLoc);
  ToStmt->setSubStmt(ToSubStmt);
  return ToStmt;
}
```
The `tie` and variable names can be eliminated if `std::get` is used but this is error prone (it is easy to use the wrong index).

Another thing: There can be an `import` for these things that are now handled by differently named functions (`ImportArrayChecked`, `ImportTemplateArgument`, `ImportTemplateArgumentListInfo` ...) so these types can be used similarly as `Stmt *`,`Decl *`, `SourceLocation *` ... in the new import template.

balazske commented:
Another (not the nicest) variant:
```
template <class T, class PreviousT>
Expected<T> import(const T &From, Expected<PreviousT> &PreviousResult) {
  if (!PreviousResult)
    return PreviousResult.takeError();
  return import(From);
}
```
```
ExpectedStmt ASTNodeImporter::VisitCaseStmt(CaseStmt *S) {
  ExpectedExpr ToLHS = import(S->getLHS());
  ExpectedExpr ToRHS = import(S->getRHS(), ToLHS);
  ExpectedStmt ToSubStmt = import(S->getSubStmt(), ToRHS);
  ExpectedSLoc ToCaseLoc = import(S->getCaseLoc(), ToSubStmt);
  ExpectedSLoc ToEllipsisLoc = import(S->getEllipsisLoc(), ToCaseLoc);
  ExpectedSLoc ToColonLoc = import(S->getColonLoc(), ToEllipsisLoc);
  if (!ToColonLoc)
    return ToColonLoc.takeError();

  CaseStmt *ToStmt = new (Importer.getToContext()) CaseStmt(
      *ToLHS, *ToRHS, *ToCaseLoc, *ToEllipsisLoc, *ToColonLoc);
  ToStmt->setSubStmt(*ToSubStmt);
  return ToStmt;
}
```

balazske committed:
Updated Stmt import functions, added generic import functions.

martong commented:
The version which returns with an `Expected<tuple<...>>` seems very reasonable and logical and I think that could indeed simplify things. Keep on going with that. 
The chained version seems a bit harder to understand and less cleaner for me.

balazske commented:
Added a new commit: New import "helper" functions added, these can be used instead of the old ones, but the old ones are left. Rewrite of the `Stmt` import functions is finished.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -95,7 +98,46 @@ namespace clang {
+      Expected<std::tuple<T>> ToOrErr = importSeq(From);
+      if (!ToOrErr)
+        return ToOrErr.takeError();
+      Expected<std::tuple<TArgs...>> To1OrErr = importSeq(FromArgs...);
```
Perhaps a better name would be `ToOrErrTail`.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -95,7 +98,46 @@ namespace clang {
+    template <class T, class ... TArgs>
+    Expected<std::tuple<T, TArgs...>>
+    importSeq(const T &From, const TArgs &...FromArgs) {
+      Expected<std::tuple<T>> ToOrErr = importSeq(From);
```
Perhaps a better name would be `ToOrErrHead`.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -5582,208 +5816,194 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
-      CaseStmt(ToLHS, ToRHS,
-               *ToCaseLocOrErr, *ToEllipsisLocOrErr, *ToColonLocOrErr);
+ExpectedStmt ASTNodeImporter::VisitCaseStmt(CaseStmt *S) {
+  auto Imp = importSeq(
```
`Imp` -> `Imported` ?

martong reviewed (commented):
Looks good.

balazske committed:
Changed visit functions of Expr.

martong reviewed (commented):
Looks good, I like the simplifications caused by `importSeq`. :+1:

balazske committed:
Updated the type import functions.

martong reviewed (commented):
:+1:

balazske committed:
Changed import functions of Decl *.

martong commented:
Just enabled CI again.

balazske committed:
Updated signature of Import functions.

Now every Import returns Expected (except the ones that never fail:
IdentifierInfo* and Selector).
Error handling is updated at the client code, mostly the error is ignored
(as it was before the change too).

balazske commented:
New commit added: Everything is updated to the new error handling, all clang tests should pass. `ASTImporter` code almost finished, missing is rename of `ImportXxxCheckedNew` to `ImportXxxChecked`.

balazske commented:
How to update the lldb code?

balazske commented:
LLDB update:
https://github.com/Ericsson/lldb/pull/1

martong commented:
run tests

balazske committed:
small corrections and function rename

balazske committed:
ASTImporter problem fixes

martong commented:
run tests

balazske committed:
remove failed imported object from ASTImporter ImportedDecls mapping

balazske committed:
Updated lldb-commit.

balazske committed:
Using ASTNodeImporter::import instead of Importer.Import where applicable.

balazske committed:
Bug fix in ASTNodeImporter::ImportOverrides.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -6084,11 +6076,11 @@ ExpectedStmt ASTNodeImporter::VisitVAArgExpr(VAArgExpr *E) {
     return TypeOrErr.takeError();
 
-  ExpectedSLoc LocStartOrErr = Importer.Import(E->getLocStart());
+  ExpectedSLoc LocStartOrErr = import(E->getLocStart());
```
What is the difference between `Importer.Import` and `import`? Is it just syntactic sugar?

balazske replied:
It is mainly syntactic simplification, but a common place for the import code too. If only `import` is used it is more easy to change it (this is used at the opensource patch when `Importer.Import` does not yet return `Expected`).

martong reviewed (commented):
Looks good.

balazske commented:
Opensource review (of a first part of these changes) is here: https://reviews.llvm.org/D51633
It is based on an older version of `ASTImporter` but update is possible and error fixes, at least it is under review now.

balazske committed:
VisitVarTemplateSpecializationDecl PointOfInstantiation fix

balazske committed:
Merge branch 'ctu-clang6' into ODR-failure

Conflicts:
	include/clang/AST/ASTImporter.h
	lib/AST/ASTImporter.cpp
	unittests/AST/ASTImporterTest.cpp

balazske committed:
SourceLocation import fix in VisitCXXDependentScopeMemberExpr

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -1007,7 +1015,8 @@ TEST_P(ASTImporterTestBase, ImportRecordTypeInFunc) {
       ImportType(FromVar->getType().getCanonicalType(), FromVar, Lang_C);
-  EXPECT_FALSE(ToType.isNull());
+  llvm::Error Err = ToType.takeError();
+  EXPECT_FALSE(Err.operator bool());
```
Do we really have to call `operator bool` explicitly?

balazske replied:
It is not needed.

balazske committed:
Small non-functional-changes, and added a missing consumeError.

balazske commented:
Error detection is improved, this results in more failing imports. The previous `ImportDefinition` or `ImportContext` like functions did not always check the errors. The problem is that in some cases an already created but not completely imported node can fail and remain in the AST. These can cause later various other asserts or crashes. A solution is to reverse the already performed operations on the AST if an import fails, but this is difficult to implement. Or collect every failed node and remove these at a later phase (but how to remove all references?). The fastest thing that looks usable: Use `setInvalidDecl` on the failed Decls. But this does not work, later function calls expect that the Decl is not invalid.

balazske commented:
To do: Some visit functions can be improved to not leave non-complete Decl's if import of something fails after the new Decl was created. Many things can be imported before the new object is created (that are now imported after it). It would be ideal to have no import at all after `GetImportedOrCreateDecl` in a visit function, but not always possible.

balazske commented:
Another possible problem: If `ImportDefinition` for `RecordDecl` fails, it can leave the RecordDecl in a isBeingDefined state. In this state the structural equivalence check will return true when the record is compared to any other (after preceeding tests). If a record with same name is imported later, we can find the previous (partly imported) record as a definition for the currently imported.

martong commented:
> Another possible problem: If ImportDefinition for RecordDecl fails, it can leave the RecordDecl in a isBeingDefined state. In this state the structural equivalence check will return true when the record is compared to any other (after preceeding tests). If a record with same name is imported later, we can find the previous (partly imported) record as a definition for the currently imported.

Sounds reasonable to me. But synthesizing tests for this improvement will be tricky.

> To do: Some visit functions can be improved to not leave non-complete Decl's if import of something fails after the new Decl was created. Many things can be imported before the new object is created (that are now imported after it). It would be ideal to have no import at all after GetImportedOrCreateDecl in a visit function, but not always possible.

Agree.

balazske committed:
Update with #490 ("Port back D51597").

balazske commented:
Diff for the new changes to improve static function importing:
https://github.com/balazske/clang/pull/1
(This should appear later as a new commit here.)

balazske committed:
Corrected import of static and anonymous functions.

balazske commented:
This patch is for ctu-clang6. Maybe it is better to wait until the https://reviews.llvm.org/D51633 and related patches are committed, then put the change into ctu-clang7.

martong commented:
> This patch is for ctu-clang6. Maybe it is better to wait until the https://reviews.llvm.org/D51633 and related patches are committed, then put the change into ctu-clang7.

I agree, we should put this into ctu-clang7.

martong commented:
ctu-clang6 is abandoned so, this is obsolete too.
